// @nemtus/symbol-sdk-typescript/cjs/symbol/Network.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var ByteArray = require('../ByteArray').ByteArray;
var Hash256 = require('../CryptoTypes').Hash256;
var BasicNetwork = require('../Network').Network;
var BasicNetworkTimestamp = require('../NetworkTimestamp').NetworkTimestamp;
var NetworkTimestampDatetimeConverter = require('../NetworkTimestamp').NetworkTimestampDatetimeConverter;
var base32 = require('../utils/base32');
var sha3_256 = require('@noble/hashes/sha3').sha3_256;
/**
 * Represents a symbol network timestamp with millisecond resolution.
 */
var NetworkTimestamp = /** @class */ (function (_super) {
    __extends(NetworkTimestamp, _super);
    function NetworkTimestamp() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Adds a specified number of milliseconds to this timestamp.
     * @param {number} count Number of milliseconds to add.
     * @returns {NetworkTimestamp} New timestamp that is the specified number of milliseconds past this timestamp.
     */
    NetworkTimestamp.prototype.addMilliseconds = function (count) {
        return new NetworkTimestamp(this.timestamp + BigInt(count));
    };
    /**
     * Adds a specified number of seconds to this timestamp.
     * @override
     * @param {number} count Number of seconds to add.
     * @returns {NetworkTimestamp} New timestamp that is the specified number of seconds past this timestamp.
     */
    NetworkTimestamp.prototype.addSeconds = function (count) {
        return this.addMilliseconds(1000n * BigInt(count));
    };
    return NetworkTimestamp;
}(BasicNetworkTimestamp));
/**
 * Represents a Symbol address.
 */
var Address = /** @class */ (function (_super) {
    __extends(Address, _super);
    /**
     * Creates a Symbol address.
     * @param {Uint8Array|string|Address} address Input string, byte array or address.
     */
    function Address(address) {
        var rawBytes = address;
        if ('string' === typeof address)
            rawBytes = base32.decode("".concat(address, "A")).slice(0, -1);
        else if (address instanceof Address)
            rawBytes = address.bytes;
        return _super.call(this, Address.SIZE, rawBytes) || this;
    }
    /**
     * Returns string representation of this object.
     * @returns {string} String representation of this object
     */
    Address.prototype.toString = function () {
        return base32.encode(new Uint8Array(__spreadArray(__spreadArray([], this.bytes, true), [0], false))).slice(0, -1);
    };
    Address.SIZE = 24;
    Address.ENCODED_SIZE = 39;
    return Address;
}(ByteArray));
/**
 * Represents a Symbol network.
 */
var Network = /** @class */ (function (_super) {
    __extends(Network, _super);
    /**
     * Creates a new network with the specified name, identifier byte and generation hash seed.
     * @param {string} name Network name.
     * @param {number} identifier Network identifier byte.
     * @param {Date} epochTime Network epoch time.
     * @param {Hash256} generationHashSeed Network generation hash seed.
     */
    function Network(name, identifier, epochTime, generationHashSeed) {
        if (generationHashSeed === void 0) { generationHashSeed = undefined; }
        var _this = _super.call(this, name, identifier, new NetworkTimestampDatetimeConverter(epochTime, 'milliseconds'), function () { return sha3_256.create(); }, function (addressWithoutChecksum, checksum) { return new Address(new Uint8Array(__spreadArray(__spreadArray([], addressWithoutChecksum, true), checksum.subarray(0, 3), true))); }, Address, NetworkTimestamp) || this;
        _this.generationHashSeed = generationHashSeed;
        return _this;
    }
    return Network;
}(BasicNetwork));
Network.MAINNET = new Network('mainnet', 0x68, new Date(Date.UTC(2021, 2, 16, 0, 6, 25)), new Hash256('57F7DA205008026C776CB6AED843393F04CD458E0AA2D9F1D5F31A402072B2D6'));
Network.TESTNET = new Network('testnet', 0x98, new Date(Date.UTC(2021, 10, 25, 14, 0, 47)), new Hash256('7FCCD304802016BEBBCD342A332F91FF1F3BB5E902988B352697BE245F48E836'));
Network.NETWORKS = [Network.MAINNET, Network.TESTNET];
module.exports = { Address: Address, Network: Network, NetworkTimestamp: NetworkTimestamp };
